### PR TITLE
(#978) HashCode scalar

### PR DESCRIPTION
**Pull request by llorllale**
For #978 
* Added new `HashCode` scalar

### CODE REVIEW

**Comment by 0crat** [comment]
Job [#1036](https://github.com/yegor256/cactoos/issues/1036) is now in scope, role is `REV`

**Comment by 0crat** [comment]
This pull request [#1036](https://github.com/yegor256/cactoos/issues/1036) is assigned to @fabriciofx[/z](https://www.0crat.com/u/fabriciofx), here is [why](https://www.0crat.com/footprint/C63314D6Z/f770f301-ea7f-4e6a-ab23-19a9e88bdf0f); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @llorllale[/z](https://www.0crat.com/u/llorllale) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be no monetary reward for this job

File: src/main/java/org/cactoos/scalar/HashCode.java [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+ */
+public final class HashCode implements Scalar<Integer> {
+    /**
+     * Hash code.
+     */
+    private final UncheckedScalar<Integer> origin;
+
+    /**
+     * Ctor.
+     * @param initial Initial value (non-zero recommended)
```
Comment (fabriciofx) [inline src/main/java/org/cactoos/scalar/HashCode.java]: @llorllale I think the recommended is a odd prime too.

File: src/main/java/org/cactoos/scalar/HashCode.java [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+     * @param multiplier Step multiplier (odd prime recommended)
+     * @param attributes The object's attributes
+     */
+    public HashCode(
```
Comment (fabriciofx) [inline src/main/java/org/cactoos/scalar/HashCode.java]: @llorllale Let's create an another ctor with a default initial and mutiplier values, something like:
```
new HashCode(final Object... attributes) {
  this(5, 67, attributes);
}
```

File: src/test/java/org/cactoos/scalar/HashCodeTest.java [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+import java.util.Objects;
+import org.hamcrest.MatcherAssert;
+import org.hamcrest.core.IsEqual;
+import org.junit.Test;
+
+/**
+ * Tests for {@link HashCode}.
+ * @since 1.0
+ */
+public final class HashCodeTest {
```
Comment (fabriciofx) [inline src/test/java/org/cactoos/scalar/HashCodeTest.java]: @llorllale Let's create a test for the new ctor as suggest above.

File: src/test/java/org/cactoos/scalar/HashCodeTest.java [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+        final Object[] attributes = {5, 31, "abc", 5, 50f, "xyz"};
+        MatcherAssert.assertThat(
+            "Value must be equal to Josh Block's implementation of hashCode()",
+            new HashCode(initial, multiplier, attributes).value(),
+            new IsEqual<>(
+                joshBloch(initial, multiplier, attributes)
+            )
+        );
+    }
+
```
Comment (fabriciofx) [inline src/test/java/org/cactoos/scalar/HashCodeTest.java]: @llorllale Let's create a test with two different objects, but with the same attributes values, must have the same hashcode.

File: src/test/java/org/cactoos/scalar/HashCodeTest.java [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+    /**
+     * Joshua Bloch's implementation of hashCode() as per Effective Java,
+     * 2nd Edition, Item 9: "Always override hashCode when you override equals".
+     * @param initial Initial value
+     * @param multiplier Step multiplier
+     * @param attributes The object's attributes
+     * @return Hashcode value
+     */
+    private static int joshBloch(
```
Comment (fabriciofx) [inline src/test/java/org/cactoos/scalar/HashCodeTest.java]: @llorllale Well, you know my and Yegor's opinion about private methods, but I put the final decision on your hands. In time: I disagree with <a href="https://en.wikipedia.org/wiki/Rule_of_three_(computer_programming)">rule of three</a>.

File: src/main/java/org/cactoos/scalar/HashCode.java (hunk 1 above)
Comment (llorllale) [reply src/main/java/org/cactoos/scalar/HashCode.java]: @fabriciofx this is what *Effective Java* says on an example of the initial value (emphasis mine):

> A nonzero initial value is used in step 1 so the hash value will be affected by
initial fields whose hash value, as computed in step 2.a, is zero. If zero were used
as the initial value in step 1, the overall hash value would be unaffected by any
such initial fields, which could increase collisions. **The value 17 is arbitrary.**

File: src/test/java/org/cactoos/scalar/HashCodeTest.java (hunk 4 above)
Comment (llorllale) [reply src/test/java/org/cactoos/scalar/HashCodeTest.java]: @fabriciofx `HashCode` does not depend on the object, it depends on whatever attributes are passed to it. The existing test covers this.

File: src/main/java/org/cactoos/scalar/HashCode.java (hunk 1 above)
Comment (fabriciofx) [reply src/main/java/org/cactoos/scalar/HashCode.java]: @llorllale The value is arbitrary but is a *big* coincidence be a prime number, right? ;) In time: "*nonzero* initial value" we shouldn't check if the value is different than zero?

File: src/test/java/org/cactoos/scalar/HashCodeTest.java (hunk 4 above)
Comment (fabriciofx) [reply src/test/java/org/cactoos/scalar/HashCodeTest.java]: @llorllale You're right. Go ahead.

File: src/main/java/org/cactoos/scalar/HashCode.java (hunk 2 above)
Comment (llorllale) [reply src/main/java/org/cactoos/scalar/HashCode.java]: @fabriciofx sure

File: src/main/java/org/cactoos/scalar/HashCode.java (hunk 1 above)
Comment (llorllale) [reply src/main/java/org/cactoos/scalar/HashCode.java]: @fabriciofx no, we shouldn't validate inputs (just like we don't have null-checks everywhere)

**Comment by llorllale** [comment]
@rultor merge

**Comment by 0crat** [comment]
Job was finished in 9 hours, bonus for fast delivery is possible (see [§36](http://www.zerocracy.com/policy.html#36))

**Comment by 0crat** [comment]
The job [#1036](https://github.com/yegor256/cactoos/issues/1036) is now out of scope

**Comment by 0crat** [comment]
Order was finished: +20 point(s) just awarded to @fabriciofx[/z](https://www.0crat.com/u/fabriciofx)

**Comment by 0crat** [comment]
Payment to `ARC` for a closed pull request, as in [§28](http://www.zerocracy.com/policy.html#28): +10 point(s) just awarded to @llorllale[/z](https://www.0crat.com/u/llorllale)

**Pull request merged**